### PR TITLE
bpo-28441: Ensure `.exe` suffix in `sys.executable` on MinGW and Cygwin

### PR DESCRIPTION
**Pull request by embray**
Based on original patch by @ma8ma 

Note: This is needed to even the run the test suite on buildbots for affected platforms; e.g.:

```
./python.exe  ./Tools/scripts/run_tests.py -j 1 -u all -W --slowest --fail-env-changed --timeout=11700 -j2
/home/embray/src/python/test-worker/3.x.test-worker/build/python -u -W default -bb -E -W error::BytesWarning -m test -r -w -j 1 -u all -W --slowest --fail-env-changed --timeout=11700 -j2
Traceback (most recent call last):
  File "./Tools/scripts/run_tests.py", line 56, in <module>
    main(sys.argv[1:])
  File "./Tools/scripts/run_tests.py", line 52, in main
    os.execv(sys.executable, args)
PermissionError: [Errno 13] Permission denied
make: *** [Makefile:1073: buildbottest] Error 1
```

https://bugs.python.org/issue28441

Possibly to do based on review of @methane :

- [x] clarify use of `./configure --with-suffix=`; possibly disable using it to override EXEEXT
  - This is still an open issue; I believe the feature currently has too much potential for misuse.  However, that should be addressed as a separate issue.
- [x] maybe add a platform-specific `#if defined(__CYGWIN__) || defined(__MINGW32__)` around the add_exe_suffix stuff
- [x] change or maybe remove note in docs for `sys.platform`
  - Because the affected platforms are dubiously supported in the first place, it's not worth mentioning in the main documentation.
- [x] change wording of news entry

### CODE REVIEW

File: Doc/library/sys.rst
Comment (ma8ma) [inline Doc/library/sys.rst]: I'd suggest adding a note which the suffix won't be appended when the path will become an invalid link (e.g. symlink can be made without the executable suffix even in cygwin).

**Review by ma8ma — COMMENTED** [review]
@embray Thanks for opening PR! I think better not skip news for the change.

File: Doc/library/sys.rst
Comment (embray) [reply Doc/library/sys.rst]: Right--I wasn't sure whether or not to include that point.  To me it seems "obvious" that that should be the case, but maybe that's worth clarifying.

As for adding a news entry I agree this change probably merits one, though it still only impacts a small number of platforms so I'll let the core team make that determination...

**Comment by embray** [comment]
Finally got around to updating this. Hopefully we can have it merged soon.

**Comment by embray** [comment]
I rebased this--it would have been good to have for 3.7.  I'm not sure if this can be done for a bugfix release.  I think it's low-enough impact on *most* platforms that it could be...

**Comment by methane** [comment]
I feel this solution is ugly...
Is there no other way other than adding ".exe" blindly?

**Comment by methane** [comment]
I found https://stackoverflow.com/a/12255332/3430118
Can't we use it?

https://github.com/mirror/boost/blob/f72119ab640b564c4b983bd457457046b52af9ee/libs/log/src/process_name.cpp#L42-L61

**Comment by embray** [comment]
This is a generic, cross-platform solution that makes perfect sense, and some version or other of this patch has already been in use for years.  I and others have just been maintaining it from Python version to Python version...

**Comment by methane** [comment]
> This is a generic, cross-platform solution

"EXE_SUFFIX is trimmed from sys.exectable" is non-generic, platform-specific issue, isn't?

**Comment by embray** [comment]
It's an issue on platforms that have `EXE_SUFFIX`.  Realistically we're talking about Windows, and *nix layers that run on Windows (Cygwin, MinGW, etc.).  But these are still different platforms.  This works across platforms that use `EXE_SUFFIX`. 

The issue here is that the `python.exe` executable will never not have the `.exe` suffix in the filename.  It's just that it can be run from the shell without manually adding the `.exe`, so the `.exe` might not be in `argv[0]`, but it should still be part of the `sys.executable` path.

File: Misc/NEWS.d/next/Library/2017-12-21-15-14-41.bpo-28441.8XZw-t.rst
Comment (methane) [inline Misc/NEWS.d/next/Library/2017-12-21-15-14-41.bpo-28441.8XZw-t.rst]: On Windows, `PC/getpathp.c` is used instead of `Module/getpath.c`.
So Windows is not affected by this change.
Cygwin and all other platforms are affected when `configure --with-suffix=.<SUFFIX>` option is used.

File: Doc/library/sys.rst
Comment (methane) [inline Doc/library/sys.rst]: EXE_SUFFIX is specified by configure option.  So "defined for the platform" is bit unclear.
Especially, suffix is added for Windows build already before this change.

**Review by methane — COMMENTED** [review]
Is there any reason to add suffix even when `isxfile(progpath)` is true without suffix?

File: Misc/NEWS.d/next/Library/2017-12-21-15-14-41.bpo-28441.8XZw-t.rst
Comment (embray) [reply Misc/NEWS.d/next/Library/2017-12-21-15-14-41.bpo-28441.8XZw-t.rst]: True, we can reword that then.

File: Doc/library/sys.rst
Comment (embray) [reply Doc/library/sys.rst]: Perhaps this note can be excluded entirely.  I see this more as a bug fix anyways: If the `.exe` is part of the executable's filename it should be included in `sys.executable` is the file's real name.

File: Misc/NEWS.d/next/Library/2017-12-21-15-14-41.bpo-28441.8XZw-t.rst
Comment (embray) [reply Misc/NEWS.d/next/Library/2017-12-21-15-14-41.bpo-28441.8XZw-t.rst]: Agreed.  Would you accept a different wording?

**Comment by embray** [comment]
> Is there any reason to add suffix even when isxfile(progpath) without suffix?

Not if the filename with the suffix doesn't exist.

There is one remote corner case where this could be incorrect, where if both `foo` and `foo.exe` both exist under the same path and are both executable.  This should never happen on a "sane" system (Cygwin will even try to prevent you from doing it, but not always successfully), but I admit it's an area where there's some inescapable ambiguity.  In that case I would err on the side of *not* adding the prefix, so that could be fixed here easily enough, despite being an unusual corner case.

What's not an unusual corner case is building/developing Python.  Not having this patch leads to a lot of confusion on case-insensitive filesystems since there's both `python.exe` and `Python/` the directory.

**Comment by methane** [comment]
> There is one remote corner case where this could be incorrect, where if both foo and foo.exe both exist under the same path and are both executable. This should never happen on a "sane" system

Since `--with-suffix` is public configure option, everyone can abuse it.
I don't think we should try adding suffix on systems like Linux.

Or should we add comment to ``--with-suffix` option like "This option should be used only platform
command is executed without suffix (e.g. Cygwin, MSYS)."?

**Comment by embray** [comment]
Perhaps, but `--with-suffix` is not just about Windows-based platforms:

```
AC_EXEEXT
AC_MSG_CHECKING(for --with-suffix)
AC_ARG_WITH(suffix,
            AS_HELP_STRING([--with-suffix=.exe], [set executable suffix]),
[
    case $withval in
    no) EXEEXT=;;
    yes)    EXEEXT=.exe;;
    *)  EXEEXT=$withval;;
    esac])
AC_MSG_RESULT($EXEEXT)

# Test whether we're running on a non-case-sensitive system, in which
# case we give a warning if no ext is given
AC_SUBST(BUILDEXEEXT)
AC_MSG_CHECKING(for case-insensitive build directory)
if test ! -d CaseSensitiveTestDir; then
mkdir CaseSensitiveTestDir
fi

if test -d casesensitivetestdir
then
    AC_MSG_RESULT(yes)
    BUILDEXEEXT=.exe
else
    AC_MSG_RESULT(no)
    BUILDEXEEXT=$EXEEXT
fi
rmdir CaseSensitiveTestDir
```

for example, I think it's also used when developing Python on OSX.

**Comment by methane** [comment]
I'm not sure about what `--with-suffix` is for too.

But we can abuse `--with-suffix`; e.g. `configure CFLAGS=-O2 --with-suffix=-o2` or `configure --with-pydebug --with-suffix=-dbg`.

So using `--with-suffix` option to determine if "platform may omit executable suffix from argv[0]" doesn't "makes perfect sense" to me.

**Comment by embray** [comment]
It was last added in https://bugs.python.org/issue230197 apparently *for* OSX in fact.

**Comment by methane** [comment]
Do you know platforms require this patch other than Cygwin and MSYS?

If this is only MSYS and Cygwin issue, how about just `#if __CYGWIN__ || __MINGW32__`.

**Comment by embray** [comment]
Perhaps it shouldn't be "abused" then, because that's not what `--with-suffix` is for.

**Comment by embray** [comment]
An earlier version of the patch did use an `#ifdef` but we decided it wasn't necessary and was maybe even harmful: https://bugs.python.org/issue28441

I would suggest instead not worrying about hypothetical misuses of `--with-suffix` (or removing/deprecating that altogether if someone is misusing it).  Consenting adults, etc.

**Comment by methane** [comment]
> It was last added in https://bugs.python.org/issue230197 apparently for OSX in fact.

Hm, BUILDEXT is added later than that.
https://github.com/python/cpython/commit/1999ef4973fb0e02872c2d970e9690061b65412d#diff-a2f8f044364f136b5879679b60c19172

**Comment by methane** [comment]
> This behavior makes sense for any platform where EXE_SUFFIX is defined (appropriately).

I don't know many platforms Python runs on.  But does this patch make sense even when
executable suffix is used but never omit?  I don't think so...

**Comment by embray** [comment]
Ok, I see.  I think, while the examples you gave like `configure CFLAGS=-O2 --with-suffix=-o2` seem like valid use cases, they're really not what this flag is meant for.  It's really only supposed to be about filename suffixes for executables, *not* appending some suffix to the python executable name itself.

There are a few possible options then:

1) Add a separate configure flag for appending a suffix to the python executable filename (similar to how `--with-suffix` might currently be misused).
2) Remove and/or deprecate `--with-suffix` altogether
3) 1) and 2)
4) Keep using `--with-suffix`, but do *not* use it to set the executable extension.  autoconf already selects the appropriate extension by seeing what the compiler outputs, so there's no good reason to override it, except in the OSX case, but  `BUILDEXEEXT` was added for that.

**Comment by embray** [comment]
If it would go down easier I'd happily add back in an `#ifdef` statement.  I just thought it would be preferable to try to avoid that (granted there are already a few for `__APPLE__` as well...

**Comment by embray** [comment]
@methane Thanks for your review--I've updated the PR description with some possible action items based on your review so far, though a few of them require clarification.  I have my own opinions about this but I'm fine with whatever.

**Comment by methane** [comment]
Sorry for late reply.  I can't approve this because I don't know all platforms on the earth Python runs on.

I prefer whitelisting platforms, to avoid making any unexpected effect happens on other unknown platforms.
So please use `#if __CYGWIN__ || __MINGW32__`. 

Otherwise,  some other core-dev who are expert of portability and minor platforms should review this.
I use only Linux and macOS.  I use MSVC build of Python.  I'm not expert of portability and minor platforms.

**Comment by embray** [comment]
@methane Thanks, no problem.  I still think the way `EXE_SUFFIX` is being used ought to be fixed, but that's a separate issue.

Originally I thought it would be better to *not* whitelist things where we can help it.  But I agree that having an `#ifdef` certainly lowers the risk for other platforms, so if that make this patch more palatable I'll certainly add it back in.

Beyond that, I'm just not sure whether or not this change merits mention in the docs (in `sys.rst`).

**Comment by methane** [comment]
Maybe, `get_program_full_path` in `PC/getpathp.c` can be copied for Cygwin and MinGW.
But I'm not sure about we can call win32 API from Cygwin and MinGW...

> Beyond that, I'm just not sure whether or not this change merits mention in the docs (in sys.rst).

Cygwin and MinGW are not first-class citizens of official Python.
So I think only NEWS entry is needed for this change.

**Comment by embray** [comment]
> Maybe, get_program_full_path in PC/getpathp.c can be copied for Cygwin and MinGW.
But I'm not sure about we can call win32 API from Cygwin and MinGW...

Yes-and-no.  Practically speaking yes, you can do it, of course.  But it runs completely counter to the spirit of those platforms, particularly in the case of Cygwin (I am less sure for MinGW), as it is supposed to be a (mostly) POSIX-compliant system, with the underlying host OS (Windows) being something of a black box.  Some projects have had a bad habit in the past of treating Cygwin just as some Windows-like system, and this can lead to all kinds of problems.  For example, the code in `PC/getpathp.c` has no idea about symlinks, or how they are represented in Cygwin.

**Comment by embray** [comment]
Ok, I cleaned this up a bit and added an `#ifdef` around the `add_exe_suffix` stuff, limiting the overall impact of this change, while improving life on those platforms that are impacted by the problem.

File: Makefile.pre.in
Comment (methane) [inline Makefile.pre.in]: This is not needed: suffix is always `.exe`.

Comment (embray) [reply Makefile.pre.in]: Indeed, now that is true.

File: Modules/getpath.c
Comment (methane) [inline Modules/getpath.c:650]: Use brackets `{}` for new code.

Comment (embray) [reply Modules/getpath.c:650]: Ah, I missed that update to PEP-7.  Just as well.  I don't like if statements without braces either.

**Comment by embray** [comment]
Would you like me to squash my commits or will github do that for us?

**Comment by methane** [comment]
No need to squash: we use "Squash and merge" button.
All commits in pull request is squashed when merging.

**Review by methane — APPROVED** [review]

**Comment by embray** [comment]
@methane Thank you for taking the time on this one!

**Comment by embray** [comment]
Is there some way I can request this change to be backported to the 3.7 branch?  Until it is, no version of Python 3.7 can build on Cygwin without this patch.

It can wait until 3.8, but I should note that #8712, which I also submitted primarily for Cygwin, has been backported.

**Comment by methane** [comment]
#8712 is clearly a bug, and it's regression.  It is bit different from this.  This changes
behavior for better MinGW / MSYS support.

@ned-deily This PR is affects only MinGW and Cygwin.  All code is guareded by `#if defined(__CYGWIN__) || defined(__MINGW32__)`.
So I believe this PR is safe to backport.  Would you accept this to be backported to 3.7?

**Comment by ned-deily** [comment]
@methane, I agree with your analysis.  One could argue that, since this is a user-visible change in behavior in a maintenance release, it's not appropriate.  On the other hand, one could argue that the change prevents failures and the risk is mitigated by being limited to non-mainstream platforms (from a cpython perspective).  Overall I'm +0 on this change so I won't oppose it.  However, if you merge it, you should add a `Notable Changes to 3.7.3` entry to the 3.7 `What's New` document.

**Comment by embray** [comment]
I also agree with the analysis regarding being a user-visible change.  That said, as @methane wrote it's still a platform-specific user-visible change, and Python does not even build on Cygwin without this patch (the downstream python package for Cygwin has almost always included a version of this patch anyways).

Should I make a PR or can it be done automatically be applying the appropriate label?

Thanks to you both.

**Comment by methane** [comment]
CIs are failing due to https://bugs.python.org/issue36131
This issue is not relating to the failure.

**Pull request merged**
